### DISCUSSION TITLE
Help with adding a filter as an option to `Paru::Pandoc.new.configure`

### DISCUSSION THREAD

**Discussion by impostorsguides** [htdebeer/paru#80] [Q&A] [answered]
Hey there, thanks for writing this library.

I'm trying to run Paru with a custom filter, but I'm running into trouble.  I've read the docs [here](https://heerdebeer.org/Software/markdown/paru/) but didn't find what I was looking for.  I saw lots of great examples on writing filters, but I didn't see anything on passing filters to the converter.  Maybe I missed it.

I've got the following code:

```
require 'paru/pandoc'
require "paru/filter"

figure_counter = 0;

image_filter = Paru::Filter.new do
  with "Image" do |image|
      figure_counter += 1
    end
end

converter = Paru::Pandoc.new

converter.configure do
  from 'markdown'
  to 'pdf'
  output 'output.pdf'
  resource_path 'assets/images'
  filter image_filter
end

converter.convert_file(ARGV[0])

puts "image count: #{figure_counter}"
```

Just a simple markdown-to-pdf converter, with a filter that counts the number of images in the markdown.  I know the options passed to `.configure` are meant to map to flags that I'd normally pass to the `pandoc` terminal command, so I took my best guess and added `filter image_filter`, since I didn't see an explicit examplr of this in the docs.

When I run this script, I get the following error:

```
$ ruby convert_my_file.rb rbenv/part-0/intro-content.md
Error running filter #<Paru::Filter:0x0000000102d0ffb8>:
Could not find executable #<Paru::Filter:0x0000000102d0ffb8>
/Users/richiethomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/paru-1.2.0/lib/paru/pandoc.rb:289:in `run_converter': error while running: (Paru::Error)

pandoc	--from=markdown \
	--to=pdf \
	--output=output.pdf \
	--resource-path=assets/images \
	--filter=\#\<Paru::Filter:0x0000000102d0ffb8\> rbenv/part-0/intro-content.md

Pandoc responded with:

Error running filter #<Paru::Filter:0x0000000102d0ffb8>:
Could not find executable #<Paru::Filter:0x0000000102d0ffb8>


	from /Users/richiethomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/paru-1.2.0/lib/paru/pandoc.rb:173:in `convert_file'
	from convert_my_file.rb:29:in `<main>'
```

It appears that Paru is expecting the `filter` arg to be "executable", but as a Ruby programmer I'm not sure that that means in this context.  Is it expecting that I put the filter code in a separate file, and pass the filename as the arg instead of the `Paru::Filter` instance itself?

Thanks in advance for any help.

EDIT:

OK I think I've made progress by extracting the filter into a file and passing a plain string representing the filter's filename:

```
# convert_my_file.rb

require 'paru/pandoc'
require 'byebug'

converter = Paru::Pandoc.new

converter.configure do
  from 'markdown'
  to 'pdf'
  output 'output.pdf'
  resource_path 'assets/images'
  filter "paru_filter.rb"
end

debugger
converter.convert_file(ARGV[0])

# paru_filter.rb

require "paru/filter"

image_filter = Paru::Filter.new do
  with "Image" do |image|
      puts "found image!"
    end
end
```

However, now I'm getting a new error:

```
$ ruby convert_my_file.rb rbenv/part-0/intro-content.md
Error running filter paru_filter.rb:
Error in $: not enough input
/Users/richiethomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/paru-1.2.0/lib/paru/pandoc.rb:289:in `run_converter': error while running: (Paru::Error)

pandoc	--from=markdown \
	--to=pdf \
	--output=output.pdf \
	--resource-path=assets/images \
	--filter=paru_filter.rb rbenv/part-0/intro-content.md

Pandoc responded with:

Error running filter paru_filter.rb:
Error in $: not enough input


	from /Users/richiethomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/paru-1.2.0/lib/paru/pandoc.rb:173:in `convert_file'
	from convert_my_file.rb:14:in `<main>'
```

The `not enough input` error is too vague for me to make sense out of.  I tried searching for `not enough input` in Paru's Github repo, but didn't see it.

**Comment by htdebeer** [comment, marked as answer]
Hi! Thanks for your issue. I see you've managed to solve most of it! To get your example working, I had to change your filter to:

```ruby
#!/bin/env ruby
require "paru/filter"

Paru::Filter.run do
  with "Image" do |image|
      puts "found image!"
    end
end
```

That first line makes it an "executable script" on my operating system. But that differs from system to system.

More importantly, you wrote `Paru::Filter.new`, but that should have been `Paru::Filter.run`. `#run` is a convenience method to create a filter, connect it to standard input and standard output, and then define and run a filter.

**Reply by impostorsguides** [reply]
Awesome, yes that indeed fixed the issue!  I had to change `puts "found image!"` to `STDERR.puts "found image!"` because it didn't like when I printed to `STDOUT`.  But it looks like I'm unblocked now, so TYSM!

> to create a filter, connect it to standard input and standard output, and then define and run a filter.

Ah, that probably explains why `input` was `nil` when I added a debugger [here](https://github.com/htdebeer/paru/blob/master/lib/paru/pandoc.rb#L272).  Makes sense.

Cheers!
